in_toto/verifylib.py

```diff
@@ -635,6 +637,113 @@ def verify_all_item_rules(items, links, target_links=None):
         link.products, target_links)
 
 
+def verify_threshold_equality(layout, chain_link_dict):
+  """
+  <Purpose>
+    Iteratively verifies threshold conditions (if a step has been
+    performed and signed by enough number of functionaries) and link
+    artifacts (materials and products of a step is same for different
+    fucntionaries) of passed dictionary (chain_link_dict).
+
+  <Arguments>
+    layout:
+            The layout specified by the project owner against which the
+            threshold will be verified.
+
+    chain_link_dict:
+            A dictionary of key-link pair with step names as keys. For each
+            step name, there are one or more keyids and corresponding
+            link objects.
+
+  <Exceptions>
+    raises an Exception if threshold is not verified.
+    ThresholdVerificationError if the step is not performed by enough
+    functionaries or if the materials and products for a step are not same
+    for all functionaries.
+
+  <Side Effects>
+    None.
+
+  """
+
+  # We are only interested in links that are related to steps defined in the
+  # Layout, so iterate over layout.steps
+  for step in layout.steps:
+    # Skip steps that don't require multiple functionaries
+    if step.threshold <= 1:
+      continue
+
+    # Extract the key_link_dict for this step from the passed chain_link_dict
+    key_link_dict = chain_link_dict[step.name]
+
+    # take one exemplary link (e.g. the first in the step_link_dict)
+    compare_link = key_link_dict.values()[0]
+
+    # Check if we have at least <threshold> links for this step
+    if len(key_link_dict) < step.threshold:
+      raise ThresholdVerificationError("Step not performed by enough functionaries!".format())
+
+    # iterate over the remaining links to compare their properties
+    for keyid, link in six.iteritems(key_link_dict):
+      keys_dict = {}
+
+      # check if the authorized functionary has signed the link
+      for key, key_link in six.iteritems(layout.keys):
+        if key == keyid:
+          keys_dict[key] = key_link
+          verify_link_signatures(link, keys_dict)
+          break
```

Comment:
You are right. I think I was confused by the variable names: `keyid`,`link`, `key`, `key_link`. :)
I think instead of iterating over `layout.keys`, you can just fetch the key with the `keyid` from the outer loop and verify the link's signature with that key, e.g.:
```python
key_dict = {
  keyid: layout.keys.get(keyid)
}
verify_link_signatures(link, key_dict)
```
And maybe you want to wrap this in a try/except block and raise a `ThresholdVerificationError` with a descriptive message if the signature verification fails.